### ISSUE TITLE
`Process::Status`: What is a _normal_ exit?

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#15231]
### `normal_exit?`

`Process::Status` has a method `normal_exit?` which "Returns `true` if the process terminated normally."
The method documentation doesn't clarify what "terminated normally" means exactly. On Unix systems it corresponds to `WIFEXITED`; on Windows it's always `true`.
Following the definition of `WIFEXITED`, "terminated normally" means "by calling exit(3) or _exit(2),  or  by returning from main()" (citing `man 2 wait` on Ubuntu).

### `ExitReason::Normal`

`Process::ExitReason::Normal` also represents a "normal" process exit. However, on Windows it does not correspond to `Status#normal_exit?`:

```cr
status = Process::Status.new(LibC::STATUS_CONTROL_C_EXIT)
status.normal_exit?      # => true
status.exit_code.normal? # => false
```

This is very irritating: Two strongly related API methods use the same term ("normal") with apparently different meanings.

The documentation for `ExitReason::Normal` further clarifies:
> * On Unix-like systems, this implies `Process::Status#normal_exit?` is true.
> * On Windows, only exit statuses less than `0x40000000` are assumed to be reserved for normal exits.

So it explicitly mentions the relation to `Status#normal_exit?`, but oddly only for Unix.

The description for Windows contradicts the observed behaviour for `#normal_exit?`: `LibC::STATUS_CONTROL_C_EXIT > 0x40000000`, so it should not be considered a normal exit.

### `signal_exit?`

The related method `#signal_exit?` returns (mostly[^1]) the inverse `#normal_exit`. It's documentation describes it as "Returns `true` if the process was terminated by a signal." On Unix systems that corresponds to `WIFSIGNALED`; on Windows it's always `false`.

Both methods are well defined, but they're only meaningful on Unix systems.
So I'm also wondering about the general usefulnes. `#signal_exit?` is clearly a very platform-specific. It cannot be used on Windows, but the concepts expressed via signals are also present on Windows, just using a different mechanism. `#signal_exit?` exposes an implementation detail of the Unix interface.
Perhaps we could replace it with a portable alternative that relates to the general concept of an abnormal exit instead of tying it to the signal mechanism.

### Conclusion

We need a clear, unique definition of what is considered a "normal exit".
That means we must either align `normal_exit?` and `ExitReason::Normal`, or rename at least on of them.

The definition of `ExitReason::Normal` makes a lot of sense to me, and its implementation even on Windows corrsponds to the definition of "normal termination" on Unix.

So I think we should adjust `#normal_exit?` to be equivalent to `exit_reason.normal?`. It would essentially be just a convenience short cut, but it certainly makes sense to have that in the API, not just for backwards compatibility.
This would of course consitute a breaking change, but only on Windows which is still in an unsupported state.

### Extra

Some additional enhancements for consideration:

* Deprecate `#signal_exit?`
* Introduce a new method `#abnormal_exit?` with an implementation equivalent to `!normal_exit?` as convenient replacement for `#signal_exit?`. The behaviour would be identical on Unix, but also reflect the corresponding concepts on Windows.
* Introduce a new method `ExitReason#abnormal?` with an implementation equivalent to `!normal?` which is a convenient summary for all reasons that are not `Normal`.

[^1]: The only exception is exit status `0x7f` which is neither `normal_exit?` nor `signal_exit?`.

**Comment by straight-shoota** [comment]
I looked a bit more into the odd case of the exit status `0x7f` on Unix, which `Process::Status` consideres neither a normal nor a signal exit.

This is in accordance with the glibc definition of WSIGNALED. But in musl libc `WSIGNALED(0x7f)` is true. So there's some differences.
Apparently treating 0x7f not as a signal exit is a Unix legacy and not actually correct on all systems. On MIPS architecture it actually seems to be a signal.[^1](https://www.openwall.com/lists/musl/2018/08/10/3) Not sure if glibc headers are different for MIPS.
The musl implementation seems to simplify things a bit while still being correct (arguably, even _more_ correct because of MIPS semantics).

I suppose it is another argument for removing `signal_exit` that we don't have to worry about such details.

**Comment by ysbaddaden** [comment]
Makes sense, including the bonus :+1:
Let's follow musl on `0x7f` (consider a signal) :+1: 

Note that the documentation uses the word `abnormal` to define `ExitReason::Aborted` ... so its description should be fixed, maybe:

> The process exited due to an abort request.

**Comment by straight-shoota** [comment]
Closed by  #15255
